lib/membrane_h264_plugin/parser.ex

```diff
@@ -269,41 +423,117 @@ defmodule Membrane.H264.Parser do
       }
 
       if Enum.any?(au, &(&1.status == :error)) or state.skip_until_keyframe do
+        Membrane.Logger.debug("skipping")
         {[], state}
       else
         {[au], state}
       end
     end)
   end
 
-  defp prepare_actions_for_aus(aus, state) do
+  @spec prepare_actions_for_aus(
+          [AUSplitter.access_unit()],
+          CallbackContext.t(),
+          state()
+        ) :: callback_return()
+  defp prepare_actions_for_aus(aus, ctx, state) do
     Enum.flat_map_reduce(aus, state, fn au, state ->
-      {sps_actions, state} = maybe_parse_sps(au, state)
-
-      au = maybe_add_parameter_sets(au, state) |> delete_duplicate_parameter_sets()
-      state = cache_parameter_sets(state, au)
+      {au, stream_format_actions, state} = process_au_parameter_sets(au, ctx, state)
 
       {{pts, dts}, state} = prepare_timestamps(au, state)
 
       buffers_actions = [
-        {:buffer, {:output, wrap_into_buffer(au, pts, dts, state.output_alignment)}}
+        buffer:
+          {:output,
+           wrap_into_buffer(au, pts, dts, state.output_alignment, state.output_stream_structure)}
       ]
 
-      {sps_actions ++ buffers_actions, state}
+      {stream_format_actions ++ buffers_actions, state}
     end)
   end
 
-  defp maybe_parse_sps(au, state) do
-    case Enum.find(au, &(&1.type == :sps)) do
-      nil ->
-        {[], state}
+  @spec process_new_parameter_sets([NALu.t()], [NALu.t()], CallbackContext.t(), state()) ::
+          {[Action.t()], state()}
+  defp process_new_parameter_sets(new_spss, new_ppss, context, state) do
+    updated_cached_spss =
+      new_spss
+      |> Enum.map(&{&1.parsed_fields.seq_parameter_set_id, &1})
+      |> Map.new()
+      |> then(&Map.merge(state.cached_spss, &1))
+
+    updated_cached_ppss =
+      new_ppss
+      |> Enum.map(&{&1.parsed_fields.pic_parameter_set_id, &1})
+      |> Map.new()
+      |> then(&Map.merge(state.cached_ppss, &1))
```

Comment:
we can consider wrapping that update logic into a separate function, so that not to repeat code